Python3 upgrade rewrite

sblack-usu opened the pull request:
I rewrote the python3 upgrade commits into two commits and cleaned up unnecessary chmods and CLRF line endings that creeped into the original python3-upgrade branch.

The first commit runs the python module 2to3 against develop.  This has a lot of changes that are automated.  The second commit squashes all of my manual commits into one, as well as removes all chmods and CLRF file changes that are present in the original python3-upgrade branch.  This helps us preserve the git history and also makes the review more digestable.  When you're reviewing, I suggest you focus on the second commit.

### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [x] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [x] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [x] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case

1. deployed to https://www.beta.hydroshare.org/
2. hs_docker_base is here https://github.com/hydroshare/hs_docker_base/pull/34

alvacouch commented on hs_core/tasks.py:
```diff
@@ -1,603 +1,601 @@
+            try:
+
+                resource = utils.get_resource_by_shortkey(res_id)
+                response_content = json.loads(response.content)
```
need encode.

sblack-usu replied:
you mean decode?

alvacouch commented on hs_core/tasks.py:
```diff
@@ -1,603 +1,601 @@
+    # get total members
+    response = session.get(url.format(list_id=list_id), auth=requests.auth.HTTPBasicAuth(
+        'hs-celery', settings.MAILCHIMP_PASSWORD))
+    total_items = json.loads(response.content)["total_items"]
```
need encode?

alvacouch commented on hs_core/tasks.py:
```diff
@@ -1,603 +1,601 @@
+                                                            settings.MAILCHIMP_PASSWORD))
+    # clear the email list
+    delete_count = 0
+    for member in json.loads(response.content)["members"]:
```
need encode?

alvacouch commented on hs_file_types/models/geofeature.py:
```diff
@@ -157,7 +157,7 @@ def get_xml(self, pretty_print=True):
 
         return CoreMetaData.XML_HEADER + '\n' + etree.tostring(RDF_ROOT, encoding='UTF-8',
-                                                               pretty_print=pretty_print)
+                                                               pretty_print=pretty_print).decode()
```
This code is a little weird, in the sense that it first encodes to utf-8 and THEN re-decodes to Unicode, before -- in subsequent code -- html-encoding the result into a file!

alvacouch commented on hs_geographic_feature_resource/models.py:
```diff
@@ -345,4 +345,4 @@ def get_xml(self, pretty_print=True, include_format_elements=True):
             self.originalcoverage.add_to_xml_container(container)
 
-        return etree.tostring(RDF_ROOT, pretty_print=pretty_print)
+        return etree.tostring(RDF_ROOT, pretty_print=pretty_print).decode()
```
not sure the default encoding is utf-8.

alvacouch commented on hs_model_program/models.py:
```diff
@@ -202,7 +202,7 @@ def get_xml(self, pretty_print=True, include_format_elements=True):
             model_code_repository.text = self.program.modelCodeRepository
 
-        xml_string = etree.tostring(RDF_ROOT, pretty_print=pretty_print)
+        xml_string = etree.tostring(RDF_ROOT, pretty_print=pretty_print).decode()
```
Same comment. I think the default is HTML/XML encoding, not utf-8 encoding.

alvacouch commented on hs_tracking/management/commands/stats.py:
```diff
@@ -197,12 +197,12 @@ def yesterdays_variables(self, lookback=1):
-                      'user_id=%s' % unicode(uid).encode(),
-                      'session_id=%s' % unicode(v.session.id).encode(),
-                      'action=%s' % unicode(v.name).encode(),
+            values = [str(v.timestamp).encode('utf-8'),
```
default is utf-8. Why do we specify it half the time, and the other half, default it?

sblack-usu replied:
There were about 10 that specified utf-8, I'll remove utf-8 when calling encode across the project

sblack-usu commented on hs_access_control/management/commands/access_permissions.py:
```diff
@@ -45,8 +45,8 @@ def handle(self, *args, **options):
             exit(1)
 
-        print(coarse_permissions(user, resource))
+        print((coarse_permissions(user, resource)))
```
That was 2to3... not sure why it put those in.

sblack-usu replied:
Looking into it, Looks like I could run 2to3 with a skip on print statements.  I think that should work fine.

https://docs.python.org/3.0/library/2to3.html#using-2to3
> When the -p is passed, 2to3 treats print as a function instead of a statement. This is useful when from __future__ import print_function is being used. If this option is not given, the print fixer will surround print calls in an extra set of parentheses because it cannot differentiate between the print statement with parentheses (such as print ("a" + "b" + "c")) and a true function call.

sblack-usu commented on hs_app_netCDF/models.py:
```diff
@@ -529,7 +529,7 @@ def get_xml(self, pretty_print=True, include_format_elements=True):
             ori_cov_obj.add_to_xml_container(container)
 
-        return etree.tostring(RDF_ROOT, pretty_print=pretty_print)
+        return etree.tostring(RDF_ROOT, pretty_print=pretty_print).decode()
```
so... probably should add an encoding parameter to etree.tostring?  Something like this?
`return etree.tostring(RDF_ROOT, pretty_print=pretty_print, encoding="UTF-8").decode()`

sblack-usu commented on hs_core/tasks.py:
```diff
@@ -1,603 +1,601 @@
-    """
-    ODM2 variables are maintained on an external site this synchronizes data to HydroShare for local caching
-    """
-    ODM2Variable.sync()
```
So. I'm open to input here.  the 2to3 tool writes the files as ASCII with CLRF endings.  After it does this, I rewrite all python files without CLRF endings.  There are, I believe 12 files that exist in the project that have CLRF endings already, this is one of them, so that's why it's showing up as a complete overwrite.  I could update it to still include CLRF endings, but I figured it makes sense to be consistent across the projects.  Since there are only 12, I found this a good time to make it consistent.

You can view the diff while ignore the whitespace change by passing w=1 as a query parameter to the diff

sblack-usu commented:
> 
> 
> There are a number of places where loads is called without decode.
> 
> ```
> ./hs_core/management/utils.py
>                              msg=str(json.loads(response.content)))
>                              msg=str(json.loads(response.content)))
> ./hs_core/tests/api/rest/base.py
>         # content = json.loads(response.content)
>         content = json.loads(response.content)
> ./hs_core/tests/api/rest/test_resource_science_meta.py
>         json.loads(js)
>         params = formclass(data=json.loads(request.body))
> ./hs_core/views/utils.py
>     data = json.loads(js_response.content)
> ./hs_core/views/__init__.py
>     total_items = json.loads(response.content)["total_items"]
>     for member in json.loads(response.content)["members"]:
>                 response_content = json.loads(response.content)
> ./hs_core/tasks.py
>     response_dict = json.loads(json_response.content)
> ./hs_file_types/views.py
>         resp_json = json.loads(resp.content)
>             resp_json = json.loads(resp.content)
>             resp_json = json.loads(resp.content)
>         resp_json = json.loads(resp.content)
>         resp_json = json.loads(resp.content)
>         resp_json = json.loads(resp.content)
>         resp_json = json.loads(resp.content)
> ./ref_ts/tests/test_ref_ts.py
>         data = json.loads(str)
> ./hs_odm2/models.py
>                 logger.debug(session_response.content)
>             logger.debug(session_response.content)
>                 logger.debug(response.content)
>             root = ElementTree.fromstring(response.content)
>                 logger.debug(response.content)
>                 for key, value in response_content["resource"].items():
>                 for url in response_content["content"]:
>         image_type = get_image_type(h=response.content)
>         base64_string = base64.b64encode(response.content)
> ./hs_tools_resource/models.py
> ```

I wasn't able to find most of these entries.  for example ./hs_core/management/utils.py does not contain any text with
>                             msg=str(json.loads(response.content)))

sblack-usu commented on hs_core/tests/api/rest/test_content_types.py:
```diff
@@ -21,7 +21,7 @@ def setUp(self):
         response = self.client.get('/hsapi/resource/content_types/', format='json')
         self.assertEqual(response.status_code, status.HTTP_200_OK)
-        content = json.loads(response.content)
+        content = json.loads(response.content.decode())
```
this is a test, so it's not going to change

sblack-usu commented:
@alvacouch - I created a new branch that does a couple of things differently.  First, it runs 2to3 against develop without running against print statements.  This cleans up a lot of clutter that's in this branch.  Then I cherry-picked the manual changes commit.  Then I made a third commit that addresses your code review comments.  Please take a look.  If the tests pass, I'll swap the close this PR and we'll start using the new one.
https://github.com/hydroshare/hydroshare/pull/3725

sblack-usu commented:
Missed a couple of things in the new branch and wanted to squash the updates. I overwrote the original and opened a new PR https://github.com/hydroshare/hydroshare/pull/3726

sblack-usu commented on django_irods/icommands.py:
```diff
@@ -15,8 +15,8 @@ def __init__(self, exitcode, stdout, stderr):
-        self.stdout = stdout
-        self.stderr = stderr
+        self.stdout = str(stdout)
+        self.stderr = str(stderr)
```
at least one test failed because stdout is already a string.  I'm rerunning it without decode or str to see if there are situations where stdout or stderr aren't strings.

sblack-usu replied:
Tests fail if I remove str() or use decode(), so I'm going to use str().  The alternative is to update usage so it's consistent... do you have a strong opinion?

sblack-usu commented on hs_app_netCDF/serialization.py:
```diff
@@ -213,7 +213,7 @@ def __str__(self):
 
         def __unicode__(self):
-            return unicode(str(self))
+            return str(str(self))
```
I've updated each occurrence of this in the other branch.

sblack-usu commented:
Check #3726 for the latest
